### PR TITLE
Kibana HTTP certificates

### LINKED ISSUE #90: Security: Kibana HTTP-side certificates
We are already mirroring the `http` attribute that we have in the Elasticsearch CRD in the Kibana CRD but no behaviour is currently associated with that structure. 

We should therefore:
* configure client/Kibana server TLS based on the `http.tls` attribute 
* default to off if not defined
* generate self-signed certificate if SAN is specified
* allow users to specify their own certificate

### PR DESCRIPTION
**Pull request by pebrc**
Fixes #90 

- [x] refactored `elasticsearch/http` package to be reused from Kibana & Elasticsearch
- [x] configure TLS in Kibana
- [x] validate TLS disabled flag in Elasticsearch
- [x] parameterize CN  in HTTP certs (Kibana/Elasticsearch)
- [x] e2e tests where only updated I was hesitant to make larger changes there as @thbkrkr refactoring is coming soon

### CODE REVIEW

File: operators/pkg/controller/common/certificates/expiration.go [hunk 1]
```diff
@@ -21,3 +21,17 @@ type RotationParams struct {
 	// RotateBefore defines how long before expiration certificates should be rotated.
 	RotateBefore time.Duration
 }
+
+// shouldRotateIn computes the duration after which a certificate rotation should be scheduled
```
Comment (thbkrkr) [inline operators/pkg/controller/common/certificates/expiration.go]: `ShouldRotateIn` with an upper case to be in sync with the code.

File: operators/pkg/controller/common/certificates/http/reconcile.go [hunk 2]
```diff
@@ -307,24 +318,41 @@ func shouldIssueNewHTTPCertificate(
...
+	cnNameParts = append(cnNameParts, "local")
+
+	var sb strings.Builder
+	for i, s := range cnNameParts {
+		if i != 0 {
+			sb.WriteString(".")
+		}
+		sb.WriteString(s) // #nosec G104
+	}
+	certCommonName := sb.String()
```
Comment (thbkrkr) [inline operators/pkg/controller/common/certificates/http/reconcile.go]: Looks like a good place to use `strings.Join`?
```go
certCommonName := strings.Join(cnNameParts, ".")
```

File: operators/test/e2e/stack/checks_k8s.go [hunk 3]
```diff
@@ -328,28 +329,34 @@ func CheckESPassword(stack Builder, k *helpers.K8sHelper) helpers.TestStep {
 func DoKibanaReq(k *helpers.K8sHelper, stack Builder, method string, uri string, body []byte) ([]byte, error) {
 	password, err := k.GetElasticPassword(stack.Elasticsearch.Name)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "while getting elastic passwor")
```
Comment (thbkrkr) [inline operators/test/e2e/stack/checks_k8s.go]: typo: password*

File: operators/pkg/controller/elasticsearch/validation/validations.go [hunk 4]
```diff
@@ -135,3 +136,14 @@ func validSanIP(ctx Context) validation.Result {
...
+
+func tlsCannotBeDisabled(ctx Context) validation.Result {
+	if !ctx.Proposed.Elasticsearch.Spec.HTTP.TLS.Enabled() {
+		return validation.Result{
+			Allowed: false,
+			Reason:  "TLS cannot be disabled for Elasticsearch currently",
+		}
+	}
+	return validation.OK
+
```
Comment (thbkrkr) [inline operators/pkg/controller/elasticsearch/validation/validations.go]: Unnecessary empty line.

File: operators/pkg/controller/common/certificates/http/reconcile.go [hunk 5]
```diff
@@ -183,11 +192,11 @@ func ensureInternalSelfSignedCertificateSecretContents(
 	}
 
 	// check if the existing cert should be re-issued
-	if shouldIssueNewHTTPCertificate(es, secret, svcs, ca, rotationParam.RotateBefore) {
+	if shouldIssueNewHTTPCertificate(owner, namer, tls, secret, svcs, ca, rotationParam.RotateBefore) {
 		log.Info(
 			"Issuing new HTTP certificate",
 			"secret", secret.Name,
-			"es", k8s.ExtractNamespacedName(&es),
+			"owner", owner,
```
Comment (thbkrkr) [inline operators/pkg/controller/common/certificates/http/reconcile.go]: We should log the `namespace` and rename `secret` in `name` or `secret_name` to be in sync with our current effort to standardize the logs (#1168).

File: operators/test/e2e/stack/pause.go [hunk 6]
```diff
@@ -1,3 +1,7 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
```
Comment (thbkrkr) [inline operators/test/e2e/stack/pause.go]: Thanks you to your IDE :)? That's the good license header. So we can remove the one below. Same in `operators/test/e2e/stack/volume.go`.

Comment (pebrc) [reply operators/test/e2e/stack/pause.go]: Actually I turned off the Intellij copyright header support and switchted to `elastic/go-licenser` and a file watch ...

**Comment by pebrc** [comment]
@thbkrkr I have fixed the issues you found could you take another look?

File: operators/pkg/apis/common/v1alpha1/common.go [hunk 7]
```diff
@@ -64,10 +64,18 @@ type TLSOptions struct {
 	Certificate SecretRef `json:"certificate,omitempty"`
 }
 
+// Enabled returns true when TLS is enabled based on this option struct.
+func (tls TLSOptions) Enabled() bool {
+	selfSigned := tls.SelfSignedCertificate
+	return selfSigned == nil || !selfSigned.Disabled || tls.Certificate.SecretName != ""
```
Comment (sebgl) [inline operators/pkg/apis/common/v1alpha1/common.go]: Looks tricky enough to deserve a unit test maybe?

Comment (pebrc) [reply operators/pkg/apis/common/v1alpha1/common.go]: added

**Pull request merged**